src/iceberg/manifest/rolling_manifest_writer.cc

```diff
@@ -0,0 +1,121 @@
+
+#include "iceberg/manifest/manifest_entry.h"
+#include "iceberg/result.h"
+#include "iceberg/util/macros.h"
+
+namespace iceberg {
+
+RollingManifestWriter::RollingManifestWriter(
+    ManifestWriterFactory manifest_writer_factory, int64_t target_file_size_in_bytes)
+    : manifest_writer_factory_(std::move(manifest_writer_factory)),
+      target_file_size_in_bytes_(target_file_size_in_bytes) {}
+
+RollingManifestWriter::~RollingManifestWriter() {
+  // Ensure we close the current writer if not already closed
+  if (!closed_) {
+    (void)CloseCurrentWriter();
+  }
+}
+
+Status RollingManifestWriter::WriteAddedEntry(
+    std::shared_ptr<DataFile> file, std::optional<int64_t> data_sequence_number) {
+  ICEBERG_ASSIGN_OR_RAISE(auto* writer, CurrentWriter());
+  ICEBERG_RETURN_UNEXPECTED(
+      writer->WriteAddedEntry(std::move(file), data_sequence_number));
+  current_file_rows_++;
+  return {};
+}
+
+Status RollingManifestWriter::WriteExistingEntry(
+    std::shared_ptr<DataFile> file, int64_t file_snapshot_id,
+    int64_t data_sequence_number, std::optional<int64_t> file_sequence_number) {
+  ICEBERG_ASSIGN_OR_RAISE(auto* writer, CurrentWriter());
+  ICEBERG_RETURN_UNEXPECTED(writer->WriteExistingEntry(
+      std::move(file), file_snapshot_id, data_sequence_number, file_sequence_number));
+  current_file_rows_++;
+  return {};
+}
+
+Status RollingManifestWriter::WriteDeletedEntry(
+    std::shared_ptr<DataFile> file, int64_t data_sequence_number,
+    std::optional<int64_t> file_sequence_number) {
+  ICEBERG_ASSIGN_OR_RAISE(auto* writer, CurrentWriter());
+  ICEBERG_RETURN_UNEXPECTED(writer->WriteDeletedEntry(
+      std::move(file), data_sequence_number, file_sequence_number));
+  current_file_rows_++;
+  return {};
+}
+
+Status RollingManifestWriter::Close() {
+  if (!closed_) {
+    ICEBERG_RETURN_UNEXPECTED(CloseCurrentWriter());
+    closed_ = true;
+  }
+  return {};
+}
+
+Result<std::vector<ManifestFile>> RollingManifestWriter::ToManifestFiles() const {
+  if (!closed_) {
+    return Invalid("Cannot get ManifestFile list from unclosed writer");
+  }
+  return manifest_files_;
+}
+
+Result<ManifestWriter*> RollingManifestWriter::CurrentWriter() {
+  if (current_writer_ == nullptr) {
+    ICEBERG_ASSIGN_OR_RAISE(current_writer_, manifest_writer_factory_());
+  } else if (ShouldRollToNewFile()) {
+    ICEBERG_RETURN_UNEXPECTED(CloseCurrentWriter());
+    ICEBERG_ASSIGN_OR_RAISE(current_writer_, manifest_writer_factory_());
+  }
+
+  return current_writer_.get();
+}
+
+bool RollingManifestWriter::ShouldRollToNewFile() const {
+  if (current_writer_ == nullptr) {
+    return false;
+  }
+  // Roll when row count is a multiple of ROWS_DIVISOR and file size >= target
+  if (current_file_rows_ % kRowsDivisor == 0) {
+    auto length_result = current_writer_->length();
+    if (length_result.has_value()) {
+      return length_result.value() >= target_file_size_in_bytes_;
+    }
+    // If we can't get the length, don't roll
```

Comment:
Not sure if this is a good idea. What if it is a fatal error returned by the underlying writer?